Towards #3041 - better Tasks in health

kolloch opened the pull request.

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala:
```diff
@@ -138,6 +139,11 @@ trait Formats
   )
 
+  implicit lazy val TaskIdFormat: Format[Task.Id] = Format(
+    Reads.of[String](Reads.minLength[String](3)).map(Task.Id(_)),
```
why 3? since we use uuid's it should be at least 36?

kolloch replied:
Admittedly, a fairly random choice. Basically, we did not validate task IDs before. Do we also want such a high length limit during testing? If the ID doesn't match an existing task, it is mood anyways.

 I think we do not really need to verify it here but my reasoning was that there needed to be at least an app prefix, a dot, and a uuid.

I can use `36` if it doesn't break any tests.

aquamatthias replied:
Is there something like nonEmpty? otherwise I don't care.

aquamatthias commented on src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala:
```diff
@@ -26,7 +26,9 @@ class TaskReplaceActor(
   val appId = app.id
-  val version = app.version.toString
+  val version = app.version
+  // FIXME: Remove!
```
Please mark as FIXME(PK) and create a ticket for you to remove all occurrences of FIXMEs.

kolloch replied:
I shouldn't have conflated this PR with changing Timestamp strings to Timestamps. I created an issue https://github.com/mesosphere/marathon/issues/3207 and I will remove the FIXME. (Otherwise I'll have to create a lot of FIXMEs ;) )

The issue contains specific instructions on how to find these occurrences. Prioritize as you like.

aquamatthias commented on src/main/scala/mesosphere/marathon/core/task/Task.scala:
```diff
@@ -3,7 +3,7 @@ package mesosphere.marathon.core.task
 import mesosphere.marathon.core.task.tracker.impl.TaskSerializer
-import mesosphere.marathon.state.{ PathId, Timestamp }
+import mesosphere.marathon.state.{ AppDefinition, PathId, Timestamp }
 import mesosphere.marathon.tasks.TaskIdUtil
```
Do we still need TaskIdUtil?

kolloch replied:
```scala
/**
  * Utility functions for dealing with TaskIDs
  */
// FIXME(PK): Remove, merge into Task.Id
class TaskIdUtil {
```

aquamatthias replied:
:+1:

aquamatthias commented on src/main/scala/mesosphere/marathon/core/task/Task.scala:
```diff
@@ -30,6 +30,16 @@ case class Task(
   def withLaunched(update: Launched => Launched): Task =
     copy(launched = launched.map(update))
 
+  def ipAddresses: Iterable[MesosProtos.NetworkInfo.IPAddress] = launched.map(_.ipAddresses).getOrElse(Iterable.empty)
```
Thanks for moving that functionality here.

kolloch replied:
:)

kolloch commented:
Thanks for the comments! I added a commit with `TaskTest` which tests the added methods -- I forgot that. `MarathonTasks.*` and `MarathonTasksTest` still exist since the code still has usages.

Since there was some doubt what constitutes to finishing #3041, I added my private checklist to this issue. Check it out and tell me if you agree.

Thanks a lot.

aquamatthias commented:
Thanks: Please squash and I will merge.

kolloch commented:
Thanks! Squashed.

aquamatthias commented:
Thanks!
